SPIRVReader changes to handle opaque pointer transition

mariusz-sikora-at-amd opened the pull request:
This patch is introducing changes to functions like:
 - postProcessRowMajorMatrix
 - addLoadInstRecursively
 - addStoreInstRecursively
 - buildConstStoreRecursively
 - transValueWithOpcode<OpLoad>
 - transValueWithOpcode<OpStore>
 - transValueWithOpcode<OpAccessChain>
 - transValueWithOpcode<OpVariable>

Additionally one workaround was added to handle OpTypeForwardPointer
this workaround will be removed when transition is competed.

amdvlk-admin commented:
Can one of the admins verify this patch?

kuhar commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -1678,29 +1718,37 @@ void SPIRVToLLVM::addStoreInstRecursively(SPIRVType *const spvType, Value *store
     for (unsigned i = 0, elementCount = spvType->getVectorComponentCount(); i < elementCount; i++) {
-      Value *const elementStorePointer =
-          getBuilder()->CreateGEP(storeType, storePointer, {zero, getBuilder()->getInt32(i)});
+      SmallVector<Value *, 2> indices;
+      indices.push_back(zero);
+      indices.push_back(getBuilder()->getInt32(i));
```
also here

kuhar commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -137,6 +137,13 @@ static void dumpLLVM(Module *m, const std::string &fName) {
   }
 }
 
+static bool isStorageClassExplicitlyLaidOut(SPIRVModule *m_bm, SPIRVStorageClassKind storageClass) {
+
```
Thanks for outlining this, I think it makes the callsites much cleaner.

nit: no need for an empty line before `return`

kuhar reviewed (commented):
Just some cosmetic suggestions from me.

Have you checked what the impact on compilation times is? Historically, we found the translation code to have a few pathologically slow edge cases that only showed up on some large shaders.

mariusz-sikora-at-amd commented:
> Just some cosmetic suggestions from me.
> 
> Have you checked what the impact on compilation times is? Historically, we found the translation code to have a few pathologically slow edge cases that only showed up on some large shaders.

Thanks for the comments. I will push new commit with changes.

I checked compilation time and the average time is  -0.1% (minus meaning faster).
The biggest slowdown is +0.91% (GeorgetownPipelineCs_0xB6464C22E9568A2A.pipe)
The biggest speedup is -1.15% (AshesPipelineVsFs_0x7B077AA1255F3655.pipe)

kuhar reviewed (dismissed):
LGTM but please get a second approval before submitting

Also, remember to squash all commits before submitting. You can force-push to update the PR or use the Squash-and-merge button.

mariusz-sikora-at-amd commented:
I squashed commits.

kosarev commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -1265,10 +1278,15 @@ bool SPIRVToLLVM::postProcessRowMajorMatrix() {
+              SmallVector<Value *, 3> indices;
+              indices.push_back(zero);
+              indices.push_back(getBuilder()->getInt32(i));
+              indices.push_back(zero);
```
We don't seem to need vectors here at all? Can we just use initialiser list or `ArrayRef<>`?

mariusz-sikora-at-amd replied:
I'm running into some issues when using ArrayRef<> and "release" version of the code. It looks like "getBuilder()->getInt32" is not initializing object properly when invoked inside initialize_list.
Do you know about any limitations for ArrayRef<>

Flakebi replied:
ArrayRef and Twine cannot be stored, so you cannot have a local variable of them (the backing memory immediately goes out of scope). It only works if you inline the list into the function call argument, like it was before.

kosarev replied:
> It looks like "getBuilder()->getInt32" is not initializing object properly when invoked inside initialize_list.

That's odd. If I read the code right, all the elements point to values stored somewhere else, so it should be safe to pass an `ArrayRef<>` or initialiser list.

mariusz-sikora-at-amd replied:
I returned to the initialization list for `CreateGEP` and added the same for `getIndexedType`. I also added comment because my intention was to have a single vector of indices and pass it to `CreateGEP` and `getIndexedType` which in my opinion is less error prone for later changes.

kosarev replied:
Just in case: you can still do it your way by declaring `indices` an ordinary array, so `Value *indices[] = ...`, and passing it as an `ArrayRef<>`. Still no need for vectors.

piotrAMD commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -333,6 +340,17 @@ Type *SPIRVToLLVM::transTypeWithOpcode<OpTypeForwardPointer>(SPIRVType *const sp
+  // it with elements) we are not able to repeat this later and get the same address of created structure.
+  // This function creates type which looks more or less like: %"type 0x7670800" = type <{ i32 }>
+  // Using transType we are able to create type which will be: <{ i32 }>
+  // Later this is causing asserts while creating GEP instructions. After transition is complited, GEP will remove
```
s/complited/completed

mariusz-sikora-at-amd replied:
done

piotrAMD commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -2804,27 +2875,42 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
     Value *newBase = base;
+    Type *newBaseType = basePointeeType;
+
+    // TODO: This assert will be removed later.
```
It would be more practical to use the same text in all the TODO's, so that it is easier to grep for all cases that need updating later (also something with "opaque" in the text would be helpful).

mariusz-sikora-at-amd replied:
done

piotrAMD commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -2840,22 +2926,34 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
+  // (llvm type). It is not possible to get this information later using transType(spvAccessChain->getType()) because if
+  // access chain is pointing to the structure of vetors/matrix then return type may contain some additional padding
+  // which will not be reflected when doing transType.
+  addAccessChainRetType(spvAccessChain, GetElementPtrInst::getIndexedType(baseEltType, indices));
```
Just something to keep in mind when you are working on other patches: this may need generalisation as I cannot convince myself that we only need to care about AccessChain as opposed to all instructions that return a pointer.

piotrAMD commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -2840,22 +2926,34 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
   }
+
+  // At this point we have to store correlation between SPIR-V AccessChain (SPIRVValue) and its dereferenced return type
+  // (llvm type). It is not possible to get this information later using transType(spvAccessChain->getType()) because if
```
Seems unfortunate that we cannot unify transType and the new map, but I do get the argument about needing a dereferenced returned type in the new code.

tsymalla commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -2830,20 +2918,33 @@ template <> Value *SPIRVToLLVM::transValueWithOpcode<OpAccessChain>(SPIRVValue *
+
+  // At this point we have to store correlation between SPIR-V AccessChain (SPIRVValue) and its dereferenced return type
+  // (llvm type). It is not possible to get this information later using transType(spvAccessChain->getType()) because if
+  // access chain is pointing to the structure of vetors/matrix then return type may contain some additional padding
```
nit: vectors

tsymalla commented on llpc/translator/lib/SPIRV/SPIRVReader.h:
```diff
@@ -272,6 +285,19 @@ class SPIRVToLLVM {
+    return nullptr;
+  }
+
+  void addAccessChainRetType(SPIRVValue *v, Type *t) {
```
This is only adding in case the SPIRVValue * was not found inside the map, so I'd suggest to call this `tryAdAccessChainRetType` or something.

mariusz-sikora-at-amd replied:
Good point. Will update

mariusz-sikora-at-amd commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -1356,10 +1372,12 @@ bool SPIRVToLLVM::postProcessRowMajorMatrix() {
-                  {getBuilder()->getInt32(0), getBuilder()->getInt32(i), getBuilder()->getInt32(0)});
-              Type *castType = pointerElem->getType()->getPointerElementType();
+              Constant *const rowIdx = getBuilder()->getInt32(i);
+              // Indices for GEP and getIndexedType must be the same.
```
Yes.
I will update the comment. You are right `GEP` can be confusing.

kuhar commented:
@mariusz-sikora-at-amd can you please close old conversations so that we know if there are outstanding issues or not? (Use the `Resolve conversation` button.)

mariusz-sikora-at-amd commented:
Pushed one more update where I added support for OpSelect (if AccessChain is a condition of the OpSelect) and I moved all workarounds to a single function `getPointeeType`.

mariusz-sikora-at-amd commented on llpc/translator/lib/SPIRV/SPIRVReader.cpp:
```diff
@@ -666,6 +679,56 @@ Type *SPIRVToLLVM::transTypeWithOpcode<OpTypeVector>(SPIRVType *const spvType, c
+    if (pointeeType)
+      return pointeeType;
+
+  } else if (isSelectOpCode(opCode)) {
```
It meant to be a functional change for OpSelect when OpAccessChain is a input and it is pointing to a matrix. But it was based on wrong assumptions about variable pointers. Now we know that it is illegal to have a variable pointer as a object of OpTypeMatrix.
This change is no longer needed. I will remove it.
Thanks !

mariusz-sikora-at-amd commented:
/rebase

mariusz-sikora-at-amd commented:
I rebased changes to trigger Jenkins.

qiaojbao commented:
retest this please

amdvlk-admin commented:
### Test summary for commit 757b0d8b945f674a5c50b09e731b8bd2a664bf8f
<details><summary>CTS tests (Failed: 0/177309)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 35753/62859 (56.9%)</li>
<li>Failed: 0/62859 (0.0%)</li>
<li>Not Supported: 27106/62859 (43.1%)</li><li>Warnings: 0/62859 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 27845/51589 (54.0%)</li>
<li>Failed: 0/51589 (0.0%)</li>
<li>Not Supported: 23744/51589 (46.0%)</li><li>Warnings: 0/51589 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 37998/62861 (60.4%)</li>
<li>Failed: 0/62861 (0.0%)</li>
<li>Not Supported: 24863/62861 (39.6%)</li><li>Warnings: 0/62861 (0.0%)</li>
</ul></details>


</ul></details>
